### PR TITLE
mpris support

### PR DESCRIPTION
**Pull request by donRaphaco**
#38

### CODE REVIEW

**Comment by bboozzoo** [comment]
Thanks for the PR, I'll review it soon.

Travis job failed on this:

```
+ set -ex
+ ./extra/fmt check
-- uncrustify version uncrustify 0.64_d
FAIL: /mnt/src/mconnect/mpris.vala (File size changed from 16312 to 16313)
PASS: /mnt/src/mconnectctl/share-iface.vala (1056 bytes)
PASS: /mnt/src/mconnectctl/device-iface.vala (1774 bytes)
PASS: /mnt/src/mconnectctl/telephony-iface.vala (981 bytes)
PASS: /mnt/src/mconnectctl/device-manager-iface.vala (1063 bytes)
PASS: /mnt/src/mconnectctl/main.vala (13669 bytes)
PASS: /mnt/src/crypt/certificate.vala (6892 bytes)
PASS: /mnt/src/mconnect/application.vala (3935 bytes)
PASS: /mnt/src/mconnect/notification.vala (4766 bytes)
PASS: /mnt/src/mconnect/transfer-download.vala (6135 bytes)
PASS: /mnt/src/mconnect/transfer-upload.vala (5837 bytes)
PASS: /mnt/src/mconnect/share.vala (6888 bytes)
PASS: /mnt/src/mconnect/telephony.vala (3988 bytes)
PASS: /mnt/src/mconnect/core.vala (4236 bytes)
PASS: /mnt/src/mconnect/mpris-proxies.vala (2137 bytes)
PASS: /mnt/src/mconnect/device-proxy.vala (7600 bytes)
PASS: /mnt/src/mconnect/property-proxy.vala (3198 bytes)
PASS: /mnt/src/mconnect/packethandlerinterface.vala (923 bytes)
PASS: /mnt/src/mconnect/logging.vala (1167 bytes)
PASS: /mnt/src/mconnect/packethandlers-proxy.vala (1636 bytes)
PASS: /mnt/src/mconnect/devicemanager-proxy.vala (5080 bytes)
PASS: /mnt/src/mconnect/ping-proxy.vala (1649 bytes)
PASS: /mnt/src/mconnect/transfer-proxy.vala (1848 bytes)
PASS: /mnt/src/mconnect/packethandlerinterface-proxy.vala (1026 bytes)
PASS: /mnt/src/mconnect/ping.vala (1578 bytes)
PASS: /mnt/src/mconnect/packethandlers.vala (2774 bytes)
PASS: /mnt/src/mconnect/share-proxy.vala (1857 bytes)
PASS: /mnt/src/mconnect/devicemanager.vala (7816 bytes)
PASS: /mnt/src/mconnect/discovereddevice.vala (3381 bytes)
PASS: /mnt/src/mconnect/devicechannel.vala (9201 bytes)
PASS: /mnt/src/mconnect/transfer-manager.vala (1818 bytes)
PASS: /mnt/src/mconnect/battery-proxy.vala (2885 bytes)
PASS: /mnt/src/mconnect/mousepad.vala (5602 bytes)
PASS: /mnt/src/mconnect/telephony-proxy.vala (1598 bytes)
PASS: /mnt/src/mconnect/transfer-manager-proxy.vala (3438 bytes)
PASS: /mnt/src/mconnect/main.vala (968 bytes)
PASS: /mnt/src/mconnect/packet.vala (6094 bytes)
PASS: /mnt/src/mconnect/battery.vala (1893 bytes)
PASS: /mnt/src/mconnect/utils.vala (7002 bytes)
PASS: /mnt/src/mconnect/transfer-interface.vala (970 bytes)
PASS: /mnt/src/mconnect/discovery.vala (3132 bytes)
PASS: /mnt/src/mconnect/io-job.vala (2391 bytes)
PASS: /mnt/src/mconnect/device.vala (19718 bytes)
PASS: /mnt/src/mconnect/config.vala (3770 bytes)
PASS: /mnt/test/mconn-crypt-vala-test.vala (4362 bytes)
PASS: /mnt/test/mconn-utils-test.vala (1726 bytes)
WARNING: code formatting check failed
+ echo 'WARNING: code formatting check failed'
```

Can you run `./extra/fmt fix`, push the updated sources? It's ok to push another commit to the branc, you don't have to fixup/amend.

File: src/mconnect/mpris.vala [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+                    mpris_player.next ();
+                    break;
+                }
+            }
+
+            if (pkt.body.has_member ("setVolume")) {
+                mpris_player.volume = (int) pkt.body.get_int_member ("setVolume") / 100.0;
+            }
+
+            if (pkt.body.has_member ("SetPosition")) {
```
Comment (bboozzoo) [inline src/mconnect/mpris.vala]: `setPosition`?

File: src/mconnect/mpris.vala [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+            bool update_needed = false;
+            Properties prop = {};
+            prop.position = mpris_player.position / 1000;
+            if (pkt.body.has_member ("Seek")) {
+                mpris_player.seek (pkt.body.get_int_member ("Seek"));
+                prop.position += pkt.body.get_int_member ("Seek") / 1000;
+                update_needed = true;
+            }
+
+            int ? volume = null;
```
Comment (bboozzoo) [inline src/mconnect/mpris.vala]: `volume` is later assigned but never used after that.

File: src/mconnect/mpris.vala [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+            if (pkt.body.has_member ("requestVolume") && pkt.body.get_boolean_member ("requestVolume")) {
+                volume = (int) (mpris_player.volume * 100.0);
+                update_needed = true;
+            }
+            if (pkt.body.has_member ("requestNowPlaying") && pkt.body.get_boolean_member ("requestNowPlaying")) {
+                if (mpris_player.metadata.contains ("xesam:title")
```
Comment (bboozzoo) [inline src/mconnect/mpris.vala]: I think you can make a little helper:
```
private static void get_metadata_string(MprisPlayerProxy player, string meta_what, out where) {
    if (mpris_player.metadata.contains (meta_what) && 
        mpris_player.metadata.get (meta_what).is_of_type (VariantType.STRING)) {
       where = mpris_player.metadata.get (meta_what).get_string();
    }
}
...
get_metadata_string(mpris_player, "xesam:artist", out pkt.title);
...
```

File: src/mconnect/mpris.vala [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+                        debug ("mpris player disconnected: %s", name);
+                        string key = "";
+                        player_list.find ((k, v) => {
+                            if (v == name) {
+                                key = k;
+                            }
+                            return v == k;
+                        });
+                        player_list.remove (key);
+                        _properties_watchers.remove (key);
```
Comment (bboozzoo) [inline src/mconnect/mpris.vala]: If I understood correctly, the mappings are:
```
player_list: player_id (eg. spotify)  -> DBus name org.mpris.MediaPlayer2.spotify
_properties_watchers: DBus name -> properties proxy
```
While here you are using player_id (eg. spotify) to remove the watcher.

File: src/mconnect/mpris.vala [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+ * Maciek Borzecki <maciek.borzecki (at] gmail.com>
+ */
+
+class MprisHandler : Object, PacketHandlerInterface {
+
+    public const string MPRIS = "kdeconnect.mpris.request";
+    public const string MPRIS_PKT = "kdeconnect.mpris";
+
+    private DBusProxy ? _dbus_watcher = null;
+    private HashTable<string, MprisPropertiesProxy ? > _properties_watchers;
```
Comment (bboozzoo) [inline src/mconnect/mpris.vala]: Can you add a comment
```
// maps player DBus name (eg. org.mpris.MediaPlayer2.spotify) to org.freedesktop.DBus.Properties
// proxy for that player
```

File: src/mconnect/mpris.vala (hunk 1 above)
Comment (donRaphaco) [reply src/mconnect/mpris.vala]: strangely "SetPosition" is correct, as well as "setVolume".

**Comment by donRaphaco** [comment]
cool thanks @bboozzoo for the detailed review.
I worked in your suggestions and enhancements.
I ran `./extra/fmt fix` before I commited (like last commit), but still the build still fails...
I'm on archlinux with `Uncrustify-0.66.1_f` looks like version `0.64_d` is more strict at some point.

**Pull request merged**